### PR TITLE
[ADD] auth_signup: New connection to user mail alert

### PR DESCRIPTION
**Pull request by Megaaaaaa**
Description of the issue/feature this PR addresses:

No way to know if someone logs into your account.

Current behavior before PR:

Prior to this, there was no way of knowing when a new device logged
into your personal account.

Desired behavior after PR is merged:

Adding the new version of the authenticate function, user's will now automaticly
receive a mail containing informations on a new connection made to their account.
This system uses a mail template sent automaticly on a new connection if the user
did not activate 2FA or if his device is not in the trusted devices of his
account except if his 2FA method is mail where no mail will be sent even if
his device is not trusted.

task-3191567

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by xmo-odoo — CHANGES_REQUESTED** [review]

File: addons/auth_signup/models/res_users.py
Comment (Megaaaaaa) [inline addons/auth_signup/models/res_users.py:4]: will be removed in next version

File: addons/auth_signup/models/res_users.py
Comment (mart-e) [inline addons/auth_signup/models/res_users.py:277]:
```suggestion
            'location':False,
```
(cf comment in the template)

File: addons/auth_signup/data/mail_template_data.xml
Comment (mart-e) [inline addons/auth_signup/data/mail_template_data.xml:383]:
```suggestion
                        <t t-if="ctx['location']">
                        Location: <t t-out="ctx['location']">City, Region, Country</t><br/>
                        </t>
```
instead of diplaying weird "Where: unknow", you can only show info when you actually have it (and d the same for the other fields too)

File: addons/auth_signup/data/mail_template_data.xml
Comment (mart-e) [inline addons/auth_signup/data/mail_template_data.xml:381]:
```suggestion
                        A new device has logged into your account. We sent you this email to verify it was indeed you.<br/><br/>
```

File: addons/auth_signup/data/mail_template_data.xml
Comment (mart-e) [inline addons/auth_signup/data/mail_template_data.xml:386]: @bouvyd do you think it's useful to add a link to the right place here?

File: addons/auth_signup/models/res_users.py
Comment (mart-e) [inline addons/auth_signup/models/res_users.py:132]:
```suggestion
                elif env.user._mfa_type() != 'email':
```

File: addons/auth_signup/models/res_users.py
Comment (mart-e) [inline addons/auth_signup/models/res_users.py:286]: if we don't know, we should not display the info instead of "unknown"

File: addons/auth_signup/data/mail_template_data.xml
Comment (bouvyd) [reply addons/auth_signup/data/mail_template_data.xml:386]: Yes, I would provide a link to the user settings for sure.

File: addons/auth_signup/data/mail_template_data.xml
Comment (bouvyd) [inline addons/auth_signup/data/mail_template_data.xml:385]: You speak of 'new device' above, i would keep the same wording here and speak of 'device' instead of 'browser'.

File: addons/auth_signup/data/mail_template_data.xml
Comment (bouvyd) [inline addons/auth_signup/data/mail_template_data.xml:381]:
```suggestion
                        Hello <t t-out="object.name or ''">Marc Demo</t>,<br/><br/>
                        This is an automated notification that a new device has been used to log into your account.<br/><br/>
```

- people logs into accounts, not devices (at least usually)
- I would avoid the 'we' that @mart-e suggested, as this is not odoo.com sending the notification but a fully automated system operated by 'we don't know who' - let's avoid talking in their name
- I would avoid the 'dear', i find it weird when automated systems try to be enticing

**Comment by bouvyd** [comment]
I agree that language should be consistant and details should be clear. However I have several objections to the content of the examples:
- I find the wording of Stripe unnecessarily worrisome ("we don't recognize the device..." implies they should always recognize the device, which is weird and wrong imho)
- Stripe does not mention the account in question (and i assume you can have more than one account in your inbox using e.g. mail forwarding or plus addressing)
- Google is lacking in details because they embed it in their 'activity' page or whatever, which we don't


I suggest this version that somewhat takes your remarks into account, as well as some of my own:
image

Improvements done:
- to-the-point wording
- all data regarding the connection is in the same place and format (not some of it inline with the text and some of it in the list)
- wording is more coherent: device and sign-in are used coherently
- change the order of actions (do nothing if it's you, otherwise change password): let's be re-assuring first for clueless users and encourage them to first think about whether it was them or not before rushing them into a password change
- we should include the wording 'near' regarding the location as GeoIP is quite vague but most users don't really know that
- avoid signature or 'Thanks' - no one actually sent this email
- order the connection information according to relevance and actionability by user: time and location first, device should be a bit more technical in some cases and IP address is the least actionable (we could frankly remove it, i doubt many people can recognize an ip address as safe, though I don't have strong feelings about that)


Ok for everybody? If you wish for changes, now is the time.

**Comment by bouvyd** [comment]
If db maintainers want to add extra actions (e.g. 'contact it security', 'contact support',etc ) they can do so but as a generic system i don't think we should include that

**Comment by bouvyd** [comment]
> What do you think about this ?

I think the problem of tasks like these with multiple stakeholders is that discussing it ad nauseam makes little sense, and I do not have it in me to challenge everything all the time - i've provided an opinion and I'll see what will stick, but in any case I will not block the PR for this (just to be clear: this is not the message of someone who's fed up with processes, etc. - I just know by experience that changing everything multiple times will clearly move into bike shedding territory fast and i'd like to avoid that - both your and my version are rather clear and actionable).

Do me one favour though: please avoid "empty space" before the colon, it's old french ^^

`via this link :` should really be `via this link:` (we had a recent PR from the community targeting all of those instances)

I leave @odony have the final word, since clearly in matter of security his opinion is usually more informed than mine.

**Review by odony — COMMENTED** [review]
Nice work, here are some remaining technical points, noted inline.

As for the design of the email, I agree with @bouvyd and would only have the following last points:
- it would be nice to have something consistent between similar emails, like the one in auth_totp_mail_enforce [^1]. You could update that one to match, either in this PR or a future one.
- still nice if you can fix the extra spaces before the colons
- the time being put inline on the first sentence separates it from the rest of the info block with the bold titles, so it's kind of lost. I would put it with the rest ;)



[^1]: https://github.com/odoo/odoo/blob/master/addons/auth_totp_mail_enforce/data/mail_template_data.xml#L4

**Comment by Megaaaaaa** [comment]
Can I have one last review ? @tde-banana-odoo @mart-e @bouvyd @xmo-odoo

**Comment by tde-banana-odoo** [comment]
> Can I have one last review ? @tde-banana-odoo @mart-e @bouvyd @xmo-odoo

After freeze is that's ok for you :) quite busy those days :o

File: addons/auth_signup/data/mail_template_data.xml
Comment (tde-banana-odoo) [inline addons/auth_signup/data/mail_template_data.xml:349]: You could use email_formatted field available on res.company model. Same for user, fallback on email_formatted.

File: addons/auth_signup/data/mail_template_data.xml
Comment (tde-banana-odoo) [inline addons/auth_signup/data/mail_template_data.xml:350]: Why not setting 'partner_to' to ``{{ object.partner_id.id }}`` ? That way we deal with partners, even if the result is the same.

File: addons/auth_signup/data/mail_template_data.xml
Comment (tde-banana-odoo) [inline addons/auth_signup/data/mail_template_data.xml:345]: :socks: if we stick to guidelines, this sould be more like "mail_template_auth_new_device_alert" (begin with model to know what is contains, then the module, then additional information)

File: addons/auth_signup/data/mail_template_data.xml
Comment (tde-banana-odoo) [inline addons/auth_signup/data/mail_template_data.xml:346]: Why "settings" in the name ?

File: addons/auth_signup/data/mail_template_data.xml
Comment (tde-banana-odoo) [inline addons/auth_signup/data/mail_template_data.xml:353]: As you use anyway an email layout, is it required to have such content ? Can't you simplify the content ? Seems there is a lot of extra layouting, better keep content as simple as possible.

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:246]: :socks: Unnecessary variable, could be one-lined

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:242]: Would say that it deserves a small comment, to help people understand globally the condition in which we consider we should alert the user (like mfa_type being something else -> ?)

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:251]: There are several used-once variables and nested if that could be simplified to ease reading I think, as you prefer (not really critical).

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:256]: Either you do a real recordset-based method (``for user in self``), either you do a model method (``@api.model`` that does not loop). But here it seems you are not sure what is your base record: self or env.user ?

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:257]: If you do a loop, can be used in a filtered (like ``for user in self.filtered('email')``

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:259]: I don't think it really adds something... if it exists it is probably the right model, creating ir model data manually that conflicts with "master" data is not really something that happens frequently, or even at all.

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:258]: What happens if the template has been removed ? Should we skip, raise, ... ? Currently it will break. Don't know what functionally you want to do.

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:266]: Why forcing False ?

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:265]: Why forcing partner ids ?

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:264]: recipient_ids could be computed by the template, so why giving it a void value, especially if we go for the 'partner_to' field usage for the template ?

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:275]: As 'format_datetime' is available in rendering context (see in mail_render_mixin), could be inlined directly in the template if you want.

Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:275]: 'date' is a bit vague (and impossible to grep if someday we want to cleanup context usage), so would have a more greppable-friendly variable like 'now_datetime' (or 'now_datetime_formatted' if you keep the format)

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:276]: location_address ?

File: addons/auth_signup/models/res_users.py
Comment (tde-banana-odoo) [inline addons/auth_signup/models/res_users.py:280]: Could be inlined directly into template to avoid too much custom context usage.

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, several small comments while reading the diff. However I wonder if we should use a mail.template, as it relies on custom context (hence hard to use in another flow), and we have to force quite a lot of values for security purpose (I suppose). Why not using directly a QWeb template and forge an email directly ?

cc @odony

File: addons/auth_signup/data/mail_template_data.xml
Comment (mart-e) [reply addons/auth_signup/data/mail_template_data.xml:250]: because it's consistent with the other emails regarding user configuration

File: addons/auth_signup/models/res_users.py
Comment (mart-e) [reply addons/auth_signup/models/res_users.py:267]: actually, does not make sence to run on more than one user

**Comment by mart-e** [comment]
@robodoo r+
@Megaaaaaa thanks for your work !

**Comment by tde-banana-odoo** [comment]
@robodoo r-

Highly suspected to break with knowledge, due to new emails being sent: https://runbot.odoo.com/runbot/build/52827293

Sorry if it is not you @mart-e , I will "buy" you a coffee in that case.

**Comment by mart-e** [comment]
I hate you ! :heart: 
Well I disabled the alert emails if totp is not configured. Will investigate later this test and see if we can spam everybody instead but at least [it passes tests in enterprise too](https://runbot.odoo.com/runbot/batch/1208207).

@robodoo r+

**Pull request closed without merging**
